### DISCUSSION TITLE
posts/how-to-write-and-run-an-ebpf-program-on-linux/

### DISCUSSION THREAD

**Discussion by giscus** [SRodi/giscus#5] [Announcements]
# posts/how-to-write-and-run-an-ebpf-program-on-linux/

Introduction eBPF (Extended Berkeley Packet Filter) is a powerful and flexible technology that allows for the execution of custom bytecode within the Linux kernel. Originally designed for packet filtering, eBPF has evolved into a general-purpose infrastructure for running sandboxed programs in the kernel space. These programs can be attached to various hooks, such as network events, system calls, and tracepoints, enabling developers to extend kernel functionality without modifying the kernel source code. This capability makes eBPF a versatile tool for performance monitoring, security enforcement, and network traffic analysis.

https://www.srodi.com/posts/how-to-write-and-run-an-ebpf-program-on-linux/

**Comment by Raguhere** [comment]
liked it, good one. am trying to write one more program, to print the packets and source server details while connecting specific ports as log file . not working

**Reply by SRodi** [reply]
Hi @Raguhere, thanks a lot for the comment, and sorry for the very late reply (I completely missed this notification for ~4 months).

For logging packets + source details on specific ports, you’ll typically want to:
* Attach an eBPF program to a TCP connect hook (e.g. `tcp_v4_connect` / `tcp_v6_connect` or the `sock/inet_sock_set_state` tracepoint).
* In the BPF program, filter on `dport` with `bpf_htons(YOUR_PORT)` and send an event to userspace via a perf buffer.
```c
if (bpf_htons(DEST_PORT) == sk->dst_port) {
    // matched
}

```
* In userspace, read those events and append them to a log file.

If it's "not working", the first thing I would do is:
* Add `bpf_printk()` in the BPF program and check sudo cat /sys/kernel/debug/tracing/trace_pipe to see if the program is actually firing.
* Check `dmesg | tail` for verifier errors or load/attach failures.

If you have the code in a repo or gist, feel free to share it and I can take a more concrete look.
